Implement new PR template to encourage better code reviews (#354)
ApollosProject/apollos-prototype: .github/PULL_REQUEST_TEMPLATE.md

## DESCRIPTION

> ### What does this PR do, or why is it needed? (land the plane)

Reviewer: The parentheses help us with deciding to adopt but seem redundant in production.

```suggestion
### What does this PR do, or why is it needed?
```

Author: I'm game for removing them on Apollos, but these parenthesis are the [core values for Differential](http://handbook.differential.com/)

> ### What design trade-offs/decisions were made? (entrepreneurial mindset)

Reviewer: Especially these. Not really sure what they mean.

```suggestion
### What design trade-offs/decisions were made?
```

> ### How do I test this PR? (ego-free communication)

Reviewer: ```suggestion
### How do I test this PR?
```

> ### What automated tests did you write? (grit, less but better, do the right thing)

Reviewer: ```suggestion
### What automated tests did you write?
```

## TODO:

- [ ] PR has a relevant title that adheres to our patch-notes/change-log style
- [ ] Closes [tag-issues-here]

> - [ ] No new warnings in tests or in-app

Reviewer 2: ```suggestion
- [ ] No new warnings in tests, in storybook, and in-app
```

> - [ ] Upload [GIF(s)](https://www.getcloudapp.com/) of iOS and Android

Reviewer 2: ```suggestion
- [ ] Upload GIF(s) of iOS and Android
```
can we remove the CloudApp advertisement?

- [ ] Set a relevant reviewer

## REVIEW:

**Manual QA**

- [ ] Manual QA on iOS and ensure it looks/behaves as expected
- [ ] Manual QA on Android and ensure it looks/behaves as expected

> **Code Review (ego-free communication, invest in people)**

Reviewer: ```suggestion
**Code Review**
```

- [ ] Read through the "Files changed" tab _very carefully_

> - [ ] Edge cases: what assumptions are made about input?

Reviewer 2: How do we use this and the checkboxes below? L29-34?

Author: As a reviewer, I think these are just things to make sure you consider in your review. Helps anyone make _better_ reviews. Tactically, I'm sure these won't always get checked off every review, but does give people a reminder during review on what's important. Did I understand your question?

Reviewer 2: What do you think about updating the title of this section to say something to the extent of
> Code Review: select >3 and expand on >1

Alternatively, if we feel our PR is becoming too Q&A heavy we could update these questions to reflect reviewer feedback. So the list could function as you were hoping and provide meaningful feedback to the PR author. For example.
> - [ ] Test coverage could be improved. Consider rewriting `FileName.tests.js`

My only concern would be that it would replace rather than supplement meaningful PR review feedback (i.e. less dialog more checkboxes). I believe that in fostering dialog we reengage and often rethink our work. I like that this list could prompt more meaningful feedback by highlighting areas of PR interest but I wouldn't want it to be a replacement for dialog.

Reviewer 2: Either way I feel like we should probably highlight L28 more prominently.

Author: The list isn't a replacement for dialogue. It's a framework for what you need to make sure you review. The code review is to check ALL the things, and conversate on as much as you can. 

L28 I feel like everyone does already, but the next set of lines actually give you a framework for _how_ to do L28. So I don't think highlighting L28 to be more important than the rest of the boxes is exactly helpful. 

I'm not sure if I understand your second suggestion 🤔 Provide a list of example feedback items?

Reviewer 2: I think the way you are intending these to function might require some tweaking to the language because they are written in the form of a question. My second suggestion reinforces your intent but simply rewords them into suggested feedback. You may be able to solve the confusion around "who are these intended for/how do I respond to these" by keeping the questions the same and changing the title to something like
> Code Review: Questions to Consider

I believe this is the best path forward. It clarifies your well meaning intent and removes much, if not all, of the confusion. Whether you leave these as checkboxes is up to you. It's a bit confusing as an interactive element but checking one could signify to the PR author that you included feedback related to that particular point.

- [ ] What kind of tests could be written?
- [ ] How might we make this easier for someone else to understand?
- [ ] Could the code be simpler?
- [ ] Will the code be easy to modify in the future?

> - [ ] What's one part of these changes that makes you excited to merge it?

Reviewer: This doesn't seem like a checkbox kinda thing. Not sure why I'd check this box as a reviewer. More of a cool thought but isn't going to change whether it's approved or not.
```suggestion

```

Author: I actually really like this as it encourages people to leave positive feedback. That's something that comes up a lot in retros about how to make PRs better - and they're better when they're not all just focused on what needs to be changed to get the PR merged

Reviewer 3: I think this one line is probably my favorite part of all of these changes. It's definitely not a "checkbox" ... it's a sentence. But knowing that someone, ANYONE else might be excited about the stuff that I've been working on is definitely an emotional boost. But maybe you're not excited about it, you just HAVE TO review this PR? It means you get to look inside and find something to be excited about. ❤️
